heat/core/manipulations.py

```diff
@@ -316,6 +329,169 @@ def concatenate(arrays, axis=0):
+    local[indices_x, indices_y] = a._DNDarray__array[indices_x]
+
+    return dndarray.DNDarray(local, gshape, a.dtype, a.split, a.device, a.comm)
+
+
+def diagonal(a, offset=0, dim1=0, dim2=1):
+    """
+    Extract a diagonal of an n-dimensional array with n > 1.
+    The returned array will be of dimension n-1.
+
+    Parameters
+    ----------
+    a: ht.DNDarray
+        The array of which the diagonal should be extracted.
+    offset: int, optional
+        The offset from the main diagonal.
+        Offset greater that zero means above the main diagonal, smaller zero is below the main diagonal.
+        Default is 0 which means the main diagonal will be selected.
+    dim1: int, optional
+        First dimension with respect to which to take the diagonal.
+        Default is 0.
+    dim2: int, optional
+        Second dimension with respect to which to take the diagonal.
+        Default is 1.
+    Returns
+    -------
+    res: ht.DNDarray
+        An array holding the extracted diagonal.
+
+    Examples
+    --------
+    >>> import heat as ht
+    >>> a = ht.array([[1, 2], [3, 4]])
+    >>> ht.diagonal(a)
+    array([1, 4])
+
+    >>> ht.diagonal(a, offset=1)
+    array([2])
+
+    >>> ht.diagonal(a, offset=-1)
+    array([3])
+
+    >>> a = ht.array([[[0, 1], [2, 3]], [[4, 5], [6, 7]]])
+    >>> ht.diagonal(a)
+    array([[0, 6],
+           [1, 7]])
+
+    >>> ht.diagonal(a, dim2=2)
+    array([[0, 5],
+           [2, 7]])
+    """
+    dim1, dim2 = stride_tricks.sanitize_axis(a.shape, (dim1, dim2))
+
+    if dim1 == dim2:
+        raise ValueError("Dim1 and dim2 need to be different")
+    if not isinstance(a, dndarray.DNDarray):
+        raise ValueError("a must be a DNDarray, got", type(a))
+    if not isinstance(offset, int):
+        raise ValueError("offset must be an integer, got", type(offset))
+
+    shape = a.gshape
+    ax1 = shape[dim1]
+    ax2 = shape[dim2]
+    # determine the number of diagonal elements that will be retrieved
+    length = min(ax1, ax2 - offset) if offset >= 0 else min(ax2, ax1 + offset)
+    # Remove dim1 and dim2 from shape and append resulting length
+    shape = tuple([x for ind, x in enumerate(shape) if ind not in (dim1, dim2)]) + (length,)
+    x, y = min(dim1, dim2), max(dim1, dim2)
+
+    if a.split is None:
+        split = None
+    elif a.split < x < y:
+        split = a.split
+    elif x < a.split < y:
+        split = a.split - 1
+    elif x < y < a.split:
+        split = a.split - 2
+    else:
+        split = len(shape) - 1
+
+    if a.split is None or a.split not in (dim1, dim2):
+        result = torch.diagonal(a._DNDarray__array, offset=offset, dim1=dim1, dim2=dim2)
+    else:
+        vz = 1 if a.split == dim1 else -1
+        off, _, _ = a.comm.chunk(a.shape, a.split)
+        result = torch.diagonal(a._DNDarray__array, offset=offset + vz * off, dim1=dim1, dim2=dim2)
+    return dndarray.DNDarray(result, shape, a.dtype, split, a.device, a.comm)
```

Comment:
should use `factories.array` instead of the DNDarray constructor. this is to handle more of the shapes and stuff like that.